### PR TITLE
thread-safe version

### PR DESCRIPTION
**Pull request by vanodevium**
#2 

As I promised, I did what I could :)

There are the brand new thread-safe constructors:
- NewSafeSet
- SafeFromSlice
- SafeFromSet

> we can change naming, no problem.

### CODE REVIEW

**Review by amit7itz — CHANGES_REQUESTED** [review]
The SafeSet is not safe enough for functions that involve multiple SafeSets (for example: Union, Equal, ...)
It's not enough to lock the current Set if you access the fields of the other Sets. Think what happens for example if you have `safeset1` and `safeset2`, and they both try to Union `safeset3` at the same time.
Ideally, you need to lock all the involved sets for every operation, but it's a bit tricky because it's very easy to cause a deadlock.

We should only provide SafeSet if we are sure it's 100% safe, until then I prefer people to use their own external lock.

**Comment by vanodevium** [comment]
@amit7itz 
Yeap. 
But all of these method was checked with `go test -race`. So, as I understand, they are safe.

Anyway, I'll try to lock every sets.

**Comment by vanodevium** [comment]
@amit7itz 

I did double check:
- new test with concurrent goroutines
- lock and unlock every SafeSet passed to methods :)

There are two options: either it is safe, or I need an test case that will show a bug

File: goset_test.go [hunk 1]
```diff
@@ -165,7 +165,7 @@ func TestSet_IsDisjoint(t *testing.T) {
 func TestSet_String(t *testing.T) {
 	s := NewSet[string]("a", "b")
 	str := fmt.Sprintf("%v", s)
-	possibleOutputs := []string{"Set[string]{a b}", "Set{b a}"}
+	possibleOutputs := []string{"Set[string]{a b}", "Set[string]{b a}", "Set{b a}"}
```
Comment (amit7itz) [inline goset_test.go]: Actually, `Set{b a}` is not a possible output, it's my mistake, you can delete it from the slice :)

**Comment by amit7itz** [comment]
@VanoDevium Assuming you lock all involved Sets, I'm actually mostly afraid of a deadlock scenario.
I'll try to build a test for it.

File: goset_test.go (hunk 1 above)
Comment (vanodevium) [reply goset_test.go]: Woohoo! Done.

**Comment by vanodevium** [comment]
@amit7itz 

Of course, build and test it please

Please note that unlocking occurs in defer mode. I don't think it will break.

But if you manage to write a test that shows a deadlock, I will definitely fix it.

**Comment by amit7itz** [comment]
@VanoDevium Here you go, a simple deadlock scenario:
```
func TestSafeSet_Equal_Deadlock_Concurrency(t *testing.T) {
	s1 := NewSafeSet[int](1, 2, 3)
	s2 := NewSafeSet[int](1, 2, 3)

	var wg sync.WaitGroup

	for i := 0; i < 10000; i++ {
		wg.Add(2)
		go func() {
			s1.Equal(s2)
			wg.Done()
		}()
		go func() {
			s2.Equal(s1)
			wg.Done()
		}()
	}

	wg.Wait()
}
```
I did it with "Equal()" but it can be done with any method that involves multiple sets.

As far as I know, the right way to avoid this is to make sure all locks are always being locked in the same order. Maybe there are other ways too 🤔

**Comment by vanodevium** [comment]
@amit7itz 
I'll dive into this...
Give me some time :)

**Comment by vanodevium** [comment]
@amit7itz 

I did my best.

Tests for deadlock are all passed. They were launched 1000000 times :)

**Comment by vanodevium** [comment]
@amit7itz 

How I love doing it!
Even though I've only known golang for a few weeks :)

**Comment by amit7itz** [comment]
Gland you enjoy :)
But know it's not concurrent safe anymore 😅 (at least Equal)
All locks must be locked during the entire operation.
If your tests didn't fail on that, it means they are not good enough 🙃

**Comment by amit7itz** [comment]
BTW, do you have a real need for a concurrent safe Set? Or you fo this just for fun & contribution?

Because I don't feel comfortable releasing such a dangerous feature without spending a few hours to test it, and I only want to spend these hours if someone is going to use it...

**Comment by vanodevium** [comment]
@amit7itz I'm not tired, but I'll cry :) 
It's a joke.

Please describe this case deeper (Equal(other) as example):
- I am getting `other` **Set** with lock\unlock
- I am running main operation with lock\unlock

Maybe I have to copy `other` **Set** with lock\unlock? 

---

I'm only doing this to contribute to the open source community.
Improve the code, add new features...

If it is useful to someone - I will be only glad.
If not, well, there's nothing to be done.

--- 
We can close MR at any time. I won't get upset!

**Comment by amit7itz** [comment]
Copying the `other` is a valid solution, but it has a performance cost. It depends on the use case whether it's better or worse than locking everything during the entire function.

---

Right now I feel like the right thing to do is not to merge this feature when there is no real demand. I also have some other approaches in mind for this problem (for example maybe to use `sync.Map`) I would like to consider.

I believe when we'll have real use cases for the concurrency problem, we'll know better what is the best solution.
I'm closing the MR for now. Please consider keeping your branch in case we want to use it in the future :) 
Thanks for your effort!

**Pull request closed without merging**